// otplib/otplib.js
/**
 * otplib
 *
 * @author Gerald Yeo <contact@fusedthought.com>
 * @version: 6.0.3
 * @license: MIT
 **/
'use strict';

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

function _interopDefault(ex) {
  return ex && (typeof ex === 'undefined' ? 'undefined' : _typeof(ex)) === 'object' && 'default' in ex ? ex['default'] : ex;
}

var hotp = _interopDefault(require('./hotp'));
var totp = _interopDefault(require('./totp'));
var authenticator = _interopDefault(require('./authenticator'));
var crypto = _interopDefault(require('crypto'));

authenticator.options = { crypto: crypto };
hotp.options = { crypto: crypto };
totp.options = { crypto: crypto };
var index = {
  Authenticator: authenticator.Authenticator,
  HOTP: hotp.HOTP,
  TOTP: totp.TOTP,
  authenticator: authenticator,
  hotp: hotp,
  totp: totp
};

module.exports = index;
